packages/eslint-plugin/rules/i18n-text-domain.js

```diff
@@ -0,0 +1,143 @@
+/**
+ * Internal dependencies
+ */
+const { TRANSLATION_FUNCTIONS } = require( '../utils' );
+
+/**
+ * Returns the text domain passed to the given translation function.
+ *
+ * @param {string} functionName Translation function name.
+ * @param {Array} args Function arguments.
+ * @return {undefined|*} Text domain argument.
+ */
+function getTextDomain( functionName, args ) {
+	switch ( functionName ) {
+		case '__':
+			return args[ 1 ];
+		case '_x':
+			return args[ 2 ];
+		case '_n':
+			return args[ 3 ];
+		case '_nx':
+			return args[ 4 ];
+		default:
+			return undefined;
+	}
+}
+
+module.exports = {
+	meta: {
+		type: 'problem',
+		schema: [
+			{
+				type: 'object',
+				properties: {
+					allowDefault: {
+						type: 'boolean',
+						default: false,
+					},
+					allowedTextDomains: {
+						type: 'array',
+						items: {
+							type: 'string',
+						},
+						uniqueItems: true,
+					},
+				},
+				additionalProperties: false,
+			},
+		],
+		messages: {
+			invalidValue: "Invalid text domain '{{ textDomain }}'",
+			invalidType: 'Text domain is not a string literal',
+			unnecessaryDefault: 'Unnecessary default text domain',
+			missing: 'Missing text domain',
+			useAllowedValue:
+				'Use one of the whitelisted text domains: {{ textDomains }}',
+		},
+		fixable: 'code',
+	},
+	create( context ) {
+		const options = context.options[ 0 ] || {};
+		const { allowDefault, allowedTextDomains = [] } = options;
+		const canFixTextDomain = allowedTextDomains.length === 1;
+
+		return {
+			CallExpression( node ) {
+				const { callee, arguments: args } = node;
+				if ( ! TRANSLATION_FUNCTIONS.includes( callee.name ) ) {
```

Comment:
I don't think we're accounting for this usage:

https://astexplorer.net/#/gist/554a136c5833496b13a172436ab01e93/9b0953e9dbfb9c260825419e80ba8a8548098d92

(`__` is at `callee.property.name`)

?

Maybe not common, but definitely valid.

Reply:
I don't necessarily consider this a blocker, consider I imagine it also affects `valid-sprintf`. Worth tracking at least.

Reply:
I don't think any of the existing rules (neither `valid-sprintf` nor the ones in `.eslintrc.js`) account for that at the moment.

I can look into it.

Reply:
Added support for this now in the new rules, see 11569f3a41768b91e7624067f3bf51ac42947599.

For `valid-sprintf` I can tackle it in #20574